pkg/server/observer.go

```diff
@@ -41,6 +41,22 @@ import (
 	"go.uber.org/zap"
 )
 
+// Observer defines the interface for observer server.
+type Observer interface {
+	observer.ObserverServer
+	Start()
+	GetEventsChannel() chan *pb.Payload
+	GetEndpointEventsChannel() chan<- monitorAPI.AgentNotify
+	GetLogRecordNotifyChannel() chan<- monitor.LogRecordNotify
+	StartMirroringIPCache(ipCacheEvents <-chan monitorAPI.AgentNotify)
+	StartMirroringServiceCache(ipCacheEvents <-chan monitorAPI.AgentNotify)
```

Comment:
How hard would it be to remove these from the interface? I would love to end up in a situation where we are able to split the observer fully into two parts/interfaces:

- One part contains a read-only reference to the ring buffer and uses it to serve the gRPC API (thus doesn't need these methods)
- The other part reads from cilium monitor to both populate the caches (when running Hubble outside of Cilium in a separate process) and the ring-buffer.

Reply:
probably not that hard. i think consumeMonitorEvents is the only thing that calls these methods:  https://github.com/cilium/hubble/commit/174a5a38b421b7c9f638e59fde8d17fdb276a185#diff-2079af2803eabcc08ae613577ea25d47R392 and local server doesn't call consumeMonitorEvents.

let me play around a bit.

Reply:
I fully agree that the observer is doing too much and should be split into smaller components with less responsibilities. However, I don't think such involved work should be part of this PR. We should maybe open an issue for that one.

Reply:
@gandro @Rolinh i opened another pr https://github.com/cilium/hubble/pull/64 . it's slightly different from what @gandro suggested, but i think it's cleaner than this pr.

- clean up the Observer interface: https://github.com/cilium/hubble/pull/64/files#diff-69945783af04950df9270c268c24c6a1R52. it adds a single method HandleMonitorSocket that encapsulates the logic for connecting to monitor socket and consuming monitor events. it's a noop for local server.
- move all the code related to monitor socket handling from cmd/serve.go to server/observer.go: https://github.com/cilium/hubble/pull/64/files#diff-69945783af04950df9270c268c24c6a1R468. cmd/serve.go simply calls HandleMonitorSocket.